usrsctplib/netinet/sctp_callout.c

```diff
@@ -151,117 +281,115 @@ sctp_os_timer_start(sctp_os_timer_t *c, uint32_t to_ticks, void (*ftn) (void *),
-	/*
-	 * Don't attempt to delete a callout that's not on the queue.
-	 */
-	if (!(c->c_flags & SCTP_CALLOUT_PENDING)) {
-		c->c_flags &= ~SCTP_CALLOUT_ACTIVE;
-		if (sctp_os_timer_current != c) {
-			SCTP_TIMERQ_UNLOCK();
-			return (0);
-		} else {
+	SCTPDBG(SCTP_DEBUG_TIMER2, "%s: now=%" PRIu32 ": request to stop callout %p\n",
+		__func__, ticks, c);
+	sctp_os_timer_cancel_impl(c);
+
+	while (c == sctp_os_timer_current) {
+		userland_thread_id_t tid;
+		sctp_userspace_thread_id(&tid);
+		if (sctp_userspace_thread_equal(tid, sctp_os_timer_current_tid)) {
 			/*
 			 * Deleting the callout from the currently running
 			 * callout from the same thread, so just return
 			 */
-			userland_thread_id_t tid;
-			sctp_userspace_thread_id(&tid);
-			if (sctp_userspace_thread_equal(tid,
-						sctp_os_timer_current_tid)) {
-				SCTP_TIMERQ_UNLOCK();
-				return (0);
-			}
-
-			/* need to wait until the callout is finished */
-			sctp_os_timer_waiting = 1;
-			wakeup_cookie = ++sctp_os_timer_wait_ctr;
-			SCTP_TIMERQ_UNLOCK();
-			SCTP_TIMERWAIT_LOCK();
-			/*
-			 * wait only if sctp_handle_tick didn't do a wakeup
-			 * in between the lock dance
-			 */
-			if (wakeup_cookie - sctp_os_timer_done_ctr > 0) {
-#if defined (__Userspace_os_Windows)
-				SleepConditionVariableCS(&sctp_os_timer_wait_cond,
-							 &sctp_os_timerwait_mtx,
-							 INFINITE);
-#else
-				pthread_cond_wait(&sctp_os_timer_wait_cond,
-						  &sctp_os_timerwait_mtx);
+			SCTPDBG(SCTP_DEBUG_TIMER2, "%s: now=%" PRIu32 ": self stop %p\n",
+				__func__, ticks, c);
+			break;
+		}
+#if defined(__Userspace__)
+		sctp_userland_cond_wait(&sctp_os_timer_current_changed, &SCTP_BASE_VAR(timer_mtx));
 #endif
-			}
-			SCTP_TIMERWAIT_UNLOCK();
+
+		if (c != sctp_os_timer_current) {
+			SCTPDBG(SCTP_DEBUG_TIMER2, "%s: now=%" PRIu32 ": waited callout completion %p\n", 
+				__func__, ticks, c);
 		}
-		return (0);
-	}
-	c->c_flags &= ~(SCTP_CALLOUT_ACTIVE | SCTP_CALLOUT_PENDING);
-	if (c == sctp_os_timer_next) {
-		sctp_os_timer_next = TAILQ_NEXT(c, tqe);
 	}
-	TAILQ_REMOVE(&SCTP_BASE_INFO(callqueue), c, tqe);
+
 	SCTP_TIMERQ_UNLOCK();
 	return (1);
 }
 
 void
 sctp_handle_tick(uint32_t elapsed_ticks)
 {
-	sctp_os_timer_t *c;
-	void (*c_func)(void *);
-	void *c_arg;
-	int wakeup_cookie;
+	static uint32_t last_heap_version_reported = 0;
 
 	SCTP_TIMERQ_LOCK();
 	/* update our tick count */
 	ticks += elapsed_ticks;
-	c = TAILQ_FIRST(&SCTP_BASE_INFO(callqueue));
-	while (c) {
-		if (SCTP_UINT32_GE(ticks, c->c_time)) {
-			sctp_os_timer_next = TAILQ_NEXT(c, tqe);
-			TAILQ_REMOVE(&SCTP_BASE_INFO(callqueue), c, tqe);
-			c_func = c->c_func;
-			c_arg = c->c_arg;
+
+	sctp_binary_heap_t* heap = &SCTP_BASE_INFO(timers_queue);
+	sctp_binary_heap_node_t* node = NULL;
+	while (0 == sctp_binary_heap_peek(heap, &node)) {
+		sctp_os_timer_t* c = (sctp_os_timer_t*)node->data;
+		if (!SCTP_UINT32_GE(ticks, c->c_time)) {
+			if (last_heap_version_reported != sctp_binary_heap_version(heap)) {
```

Comment:
Note for reviewers:
Currently `sctp_handle_tick` is invoked `~10 ms`, which could result in polluted logs saying  `the next soonest callout ...`. Remembering last version printed allow not printing same message multiple times.
The further enhancement would be changing `sctp_handle_tick` invocation to be result of timed wait on event/conditional variable (this will be done as dedicated PR once this PR is merged).